Creating a new class for lifetime cost calculation

sofiapinto opened the pull request:
---

# Description

This PR creates the LifetimeCostCalculator() class and adjusts the code for calculating costs for all archetypes.

Fixes #12 

# Instructions for Reviewer

Hey @eglucas ,

I've left instructions on what needs to be reviewed in the comments below. In general, can you make sure the changes to the calculations make sense and that the class functions output what we need?

Thanks in advance!

# Checklist:

- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

sofiapinto committed:
removing functions from cost_computations.py that didn't relate to maintenance or upfront costs

sofiapinto committed:
renaming cost_computations.py to maintenance_and_installation_costs.py

sofiapinto committed:
updating data getters to make archetype label the index

sofiapinto committed:
updating running_costs.py

sofiapinto committed:
changes to maintenance and installation costs.py

sofiapinto committed:
adding class file lifetime_cost_calculator.py

sofiapinto committed:
adding notebook

sofiapinto committed:
adding missing docs

sofiapinto commented on asf_lifetime_cost_model/analysis/notebooks/lifetime_cost_calculator_demo.py:
I need to add more examples here!

eglucas replied:
I added a boiler example!

sofiapinto commented on asf_lifetime_cost_model/getters/data_getters.py:
Looks like a lot changed here, but actually, most of it was just ruff wanting to apply changes to the existing code that had been merged (when i ran pre-commit). the main change was adding the following to a few functions to make the archetype label the index.

```
data.set_index("archetype_label", inplace=True)
    return data
```

sofiapinto commented on asf_lifetime_cost_model/analysis/notebooks/lifetime_cost_calculator_demo.py:
```diff
@@ -0,0 +1,55 @@
+print(running_costs)
+
+# %%
+calculator.compute_total_lifetime_costs(
```
this will look like it's not working properly but it is. We just need to fix the archetype labels in a separate PR.

sofiapinto commented on asf_lifetime_cost_model/pipeline/cost_computations.py:
I renamed this file to maintenance_and_installation_costs, so you can just look at that one

sofiapinto commented on asf_lifetime_cost_model/pipeline/lifetime_cost_calculator.py:
This is the main new file - I've moved some of the functions in our original cost calculation .py files to here and altered them so that the data isn't loaded every time the functions are called & so that the calculations are done to all archetypes at once.

sofiapinto replied:
Still needs documentation in functions to be improved + more docs at the top of the .py file

sofiapinto commented on asf_lifetime_cost_model/pipeline/maintenance_and_installation_costs.py:
```diff
@@ -0,0 +1,84 @@
+    return subsidy_value
+
+
+def compute_cost_with_loan(upfront_cost: float, life_span: int, loan_interest_rate: float) -> float:
```
new function! moved your original code to it's own function

sofiapinto replied:
although it says this is a new file, it's just because i renamed it, nothing here is new. just moved some functions to the new lifetime_cost_calculator.py

sofiapinto commented on asf_lifetime_cost_model/pipeline/running_costs.py:
not much changed here - most changes happened from running pre-commit. I moved some functions to the new lifetime_cost_calculator.py file

eglucas commented on asf_lifetime_cost_model/pipeline/lifetime_cost_calculator.py:
```diff
@@ -0,0 +1,354 @@
+)
+
+
+def get_installation_cost(costs_data: pd.DataFrame, heating_system: str, decile: int = None) -> pd.DataFrame:
```
Just wondering why you've chosen to have `get_installation_cost` in this file, and not in `data_getters`?

sofiapinto replied:
i didn't because it's not your typical getter. You're actually providing the cost_data as an argument. But i've now moved it to data_getters.

eglucas commented on asf_lifetime_cost_model/pipeline/lifetime_cost_calculator.py:
```diff
@@ -0,0 +1,354 @@
+    """A class to calculate the lifetime cost of heating systems, including installation, maintenance, and running costs."""
+
+    def __init__(self):
+        self.ashp_installation_costs = data_getters.get_ashp_installation_costs()
+        self.boiler_installation_costs = data_getters.get_gas_boiler_installation_costs()
+        self.ashp_subsidy_options_data = data_getters.get_ashp_subsidy_options_data()
+        self.wholesale_price_projections = data_getters.get_desnz_wholesale_price_projections()
+        self.property_heat_demand = data_getters.get_property_heat_demand()
+        self.property_archetypes = self.property_heat_demand.index.tolist()
```
Cool! I wouldn't have thought to assign the attributes to be the raw data, I think this is a really neat way of organising things.

sofiapinto replied:
I've had the idea from Liz's past work:
- https://github.com/nestauk/nlp-link/blob/main/nlp_link/linker.py
- https://github.com/nestauk/dap_prinz_green_jobs/blob/dev/dap_prinz_green_jobs/pipeline/green_measures/skills/green_skill_classifier.py

eglucas commented on asf_lifetime_cost_model/pipeline/lifetime_cost_calculator.py:
```diff
@@ -0,0 +1,354 @@
+            if life_span is None:
+                raise ValueError("Please provide `life_span` in years when purchasing with loans.")
+
+            if loan_interest_rate < 0.0 or loan_interest_rate > 1.0:
```
I don't think we need to change this because I assume we are trying to safeguard against inputs that are percentages, but just wanted to note that strictly speaking I think you can get interest rates that are over 100%. In any case, I think this is unlikely in the context of ASHP loans! Just wanted to flag this in case.

sofiapinto replied:
good shout, we can revisit in future if needed after user testing

eglucas commented on asf_lifetime_cost_model/pipeline/lifetime_cost_calculator.py:
```diff
@@ -0,0 +1,354 @@
+                    cost_time_series[year] *= 1.05
+
+        ### Compute total running cost in each year of operation ###
+        running_costs_time_series = pd.DataFrame()
```
Thank you for converting this into a dataframe!

eglucas commented on asf_lifetime_cost_model/pipeline/lifetime_cost_calculator.py:
```diff
@@ -0,0 +1,354 @@
+            pd.DataFrame: A DataFrame with property archetypes as index and total lifetime costs as column.
+        """
+        total_lifetime_costs = pd.DataFrame()
+        total_lifetime_costs["lifetime_installations"] = installation_costs
```
```suggestion
        total_lifetime_costs["lifetime_installation_costs"] = installation_costs.sum(axis=1)
```

1. Fix typo in dataframe column name
2. If we decide to implement my other suggestion on separating out loan amount, then this summation is needed

eglucas commented on asf_lifetime_cost_model/pipeline/lifetime_cost_calculator.py:
```diff
@@ -0,0 +1,354 @@
+                ),
+                axis=1,
+            )
+
```
```suggestion
            installation_costs["loan_interest"] = installation_costs.apply(
                lambda x: (
                    compute_cost_with_loan(
                        upfront_cost=x["installation_cost"],
                        life_span=life_span,
                        loan_interest_rate=loan_interest_rate,
                    )
                    - x["installation_cost"]
                ),
                axis=1,
            )

```

I suggest adding the loan interest amount as a separate column to make our life easier when we need to separate out installation cost vs. loan payment in the output figures of the app.

eglucas committed:
add boiler examples

eglucas reviewed (commented):
Amazing work @sofiapinto !! Thank you so much for doing all the hard work thinking through how to re-organise all the functions. The `LifetimeCostCalculator` works really well and I do think this is the best approach in terms of being able to handle all archetypes at once.

My main suggestion is to separate out the loan interest payment from the installation cost - have left more details below. 
(Side note: Depending on how we want to present the breakdown of lifetime costs in an output figure, the loan payment might have to have its own column in the total_lifetime_costs dataframe - but have left it aggregated with installation_costs for now.)

The demo notebook worked as expected, and I added a boiler example that I also tested.

Thank you so much again, this is so good! Once you've decided on how best to include the loan interest payment, happy to merge.

sofiapinto committed:
move function to data getters

sofiapinto committed:
fixing typo

Co-authored-by: eglucas <101257595+eglucas@users.noreply.github.com>

sofiapinto committed:
adding loan interest amount as a separate column

Co-authored-by: eglucas <101257595+eglucas@users.noreply.github.com>

sofiapinto committed:
separating subsidy, loan and installation cost into multiple columns

sofiapinto commented:
Hey @eglucas ,

I have:
- Moved `get_installation_cost()` to `data_getters.py`
- Separated the loan interest from the installation cost
- Related to the above I changed the code to have: installation cost, subsidy value and loan interest all in separate columns. We always present both the subsidy value and loan interest, even if these are 0.
- Added the function `compute_annualised_lifetime_costs()` to compute the annualised lifetime cost.

Let me know if you disagree with any of the changes or if you think the logic is wrong in any of them.

sofiapinto committed:
change to package the repo

sofiapinto committed:
Roisin's suggestion to include config when packaging

sofiapinto committed:
removing scenarios file

sofiapinto commented on pyproject.toml:
the changes to this file enable asf_lifetime_cost_model to be used as a package

sofiapinto commented on MANIFEST.in:
adding this file enables asf_lifetime_cost_model to be used as a package

sofiapinto committed:
temporary fix price cap data

eglucas commented on asf_lifetime_cost_model/getters/data_getters.py:
```diff
@@ -194,10 +194,10 @@ def get_current_energy_price_cap_tariffs(
         current price cap.
     """
-    current_price_cap_period = datetime.now().strftime("%Y-%m-%d")
+    # current_price_cap_period = datetime.now().strftime("%Y-%m-%d")
```
I know this was just a temporary fix so the more stable fix is to change the data source URLs in the config file of the `asf_levies_model` package installed. The way the tariff data getter is coded is that it gets today's date and finds the corresponding price cap period data.

As we're using `uv`, the config file is located at `asf_lifetime_cost_model/.venv/lib/python3.13/site-packages/asf_levies_model/config/base.yaml`

The URLs to change to the most recent price cap data are shown here: https://github.com/nestauk/asf_levies_model/commit/30e2cf03af4ba0c5fa728680089b00044de83908

I'm not sure if it's best to manually edit the config files, or to re-install the `asf_levies_model` package (which I'm assuming will automatically have the most recent config files according to the dev branch of that repo). Either way, every time there is a new price cap - the data source URLs need to be updated. Maybe in future we'll develop a way to automate this where our getter can automatically scan for the most recent URL but there are currently no plans to do that yet.

eglucas replied:
I've just seen that you opened an issue https://github.com/nestauk/asf_lifetime_cost_model/issues/28 for this fix! Will copy and paste what I wrote in my other comment to that issue.

eglucas commented on asf_lifetime_cost_model/pipeline/lifetime_cost_calculator.py:
```diff
@@ -0,0 +1,365 @@
+        installation_costs: pd.DataFrame,
+        maintenance_costs: pd.DataFrame,
+        running_costs: pd.DataFrame,
+        subsidy_model: Optional[str] = None,
```
```suggestion
```

Not sure if I'm missing something but it doesn't look like this argument is used in this function since the subsidy value is taken from the `installation_costs` dataframe :)

eglucas reviewed (approved):
Thanks for making those additional changes, Sofia!! Just have one minor comment on what looks like a redundant argument.

I've run the demo notebook and all works as expected. I think it will definitely make downstream processing and analysis way easier having the installation cost, subsidy and interest costs always separated - thank you!

Would love to discuss offline the changes you had to make to package the repo as it's all new to me. Am assuming it's all working since the app is live :)

Happy for everything to be merged.

eglucas committed:
fix wholesale projections modification

eglucas commented on asf_lifetime_cost_model/pipeline/running_costs.py:
```diff
@@ -43,14 +45,22 @@ def get_wholesale_price_projection_series(
+    # high fossil fuel prices scenarios
+    elif projection_scenario == "low fossil fuel prices":
+        wholesale_prices_data.loc[
+            wholesale_prices_data["projection_scenario"] == projection_scenario,
+            range(2032, 2051),
+        ] = wholesale_prices_data.loc[
+            wholesale_prices_data["projection_scenario"] == projection_scenario,
+            2031,
+        ]
+    elif projection_scenario == "high fossil fuel prices":
         wholesale_prices_data.loc[
-            :,
+            wholesale_prices_data["projection_scenario"] == projection_scenario,
             range(2032, 2051),
         ] = wholesale_prices_data.loc[
-            :,
+            wholesale_prices_data["projection_scenario"] == projection_scenario,
             2031,
         ]
```
This fixes the errors (issue https://github.com/nestauk/asf_lifetime_cost_model/issues/25) that we identified during app testing ("Cheaper electricity" preset scenario wasn't working because the way that the data was being modified for "low fossil fuel prices" and "high fossil fuel prices" wasn't correct)

eglucas committed:
update demo notebooks
